README.md
# Emotes and Speech Bubbles

Add some fanciness to multiplayer communication with in-game emotes and speech bubbles. Bring up the emote "wheel" with the mouse wheel button to use/swap emotes. Additionally, anything said in chat is displayed in a speech bubble over the player's head. 

If you have any ideas for more emotes feel free to drop a suggestion. With an image/gif link or YouTube video, it's pretty easy for me to add it in.

## Attributions
- Engineer Girl: https://www.reddit.com/user/HideBoar/

![Imgur](https://i.imgur.com/z5wiI1i.gif) ![Imgur](https://i.imgur.com/0uOLeN5.gif) ![Imgur](https://i.imgur.com/JLTybqa.gif) ![Imgur](https://i.imgur.com/dsGSgJb.gif) ![Imgur](https://i.imgur.com/z7i64F8.gif)

- Think spin: [KubeRoot](https://github.com/KubeRoot)

![Imgur](https://i.imgur.com/vi6f8eK.gif)

- Coggers: voparoS_

![Imgur](https://i.imgur.com/8titZh2.gif)

- 3D animated emojis: https://giphy.com/emoji

![Imgur](https://i.imgur.com/8PZuucp.gif) ![Imgur](https://i.imgur.com/Rm5EE7L.gif) ![Imgur](https://i.imgur.com/6OzM5gA.gif) ![Imgur](https://i.imgur.com/1vs6RXZ.gif) ![Imgur](https://i.imgur.com/hPsFrDL.gif) ![Imgur](https://i.imgur.com/yvF8YpP.gif) ![Imgur](https://i.imgur.com/eYaSXwc.gif)

- Party Parrot: https://cultofthepartyparrot.com/

![Imgur](https://i.imgur.com/IntiUKI.gif)

- Bear hug: Skype

![Imgur](https://i.imgur.com/MaINRvO.gif)

- Unknown

![Imgur](https://i.imgur.com/yc82Ejy.gif) ![Imgur](https://i.imgur.com/fmXTPv5.gif) ![Imgur](https://i.imgur.com/a5UfySO.gif) ![Imgur](https://i.imgur.com/k9upOfW.gif) ![Imgur](https://i.imgur.com/xmyGfuY.gif)

If you know owner of the other gifs I used let me know I just grabbed them off google images


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 4/25/22
  Additions:
    - Some new emotes recommended by Spazzbot
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 4/4/22
  Additions:
    - Speech bubble text color selector, default is same as player color
    - Emote duration setting
    - 3 new emotes
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 1/17/21
  Additions:
    - GUI size buttons
  Changes:
    - Changed wording from "Right Click: Select" to "Right Click: Change"
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 1/17/21
  Additions:
    - Emote "Wheel", press middle mouse button to bring up
    - Left click emote to use it, right click to swap it for a different one.


data.lua
local emotes = {
{
	type = "animation",
	name = "emote2",
	filename = "__EmotesAndSpeechBubbles__/graphics/smile.png",
	size = 200,
	frame_count = 10,
	line_length = 5,
	animation_speed = 0.5,
	scale = 0.5
},
{
	type = "animation",
	name = "emote3",
	filename = "__EmotesAndSpeechBubbles__/graphics/AAHHH.png",
	size = 200,
	frame_count = 10,
	line_length = 5,
	animation_speed = 0.5,
	scale = 0.5
},
{
	type = "animation",
	name = "emote4",
	filename = "__EmotesAndSpeechBubbles__/graphics/colonP.png",
	size = 200,
	frame_count = 10,
	line_length = 5,
	animation_speed = 0.5,
	scale = 0.5
},
{
	type = "animation",
	name = "emote5",
	filename = "__EmotesAndSpeechBubbles__/graphics/want.png",
	size = 200,
	frame_count = 10,
	line_length = 5,
	animation_speed = 0.5,
	scale = 0.5
},
{
	type = "animation",
	name = "emote6",
	filename = "__EmotesAndSpeechBubbles__/graphics/glare.png",
	size = 200,
	frame_count = 10,
	line_length = 5,
	animation_speed = 0.5,
	scale = 0.5
},
{
	type = "animation",
	name = "emote7",
	filename = "__EmotesAndSpeechBubbles__/graphics/hmm.png",
	size = 64,
	frame_count = 60,
	line_length = 5,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote8",
	filename = "__EmotesAndSpeechBubbles__/graphics/hmmspin.png",
	size = 64,
	frame_count = 30,
	line_length = 5,
	animation_speed = 0.75
},
{
	type = "animation",
	name = "emote9",
	filename = "__EmotesAndSpeechBubbles__/graphics/PointRight.png",
	size = 64,
	frame_count = 38,
	line_length = 2,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote10",
	filename = "__EmotesAndSpeechBubbles__/graphics/PointLeft.png",
	size = 64,
	frame_count = 38,
	line_length = 2,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote11",
	filename = "__EmotesAndSpeechBubbles__/graphics/OkHand.png",
	size = 64,
	frame_count = 60,
	line_length = 5,
	animation_speed = 0.5,
},
{
	type = "animation",
	name = "emote12",
	filename = "__EmotesAndSpeechBubbles__/graphics/clap.png",
	size = 64,
	frame_count = 24,
	line_length = 4,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote13",
	filename = "__EmotesAndSpeechBubbles__/graphics/laugh.png",
	size = 64,
	frame_count = 24,
	line_length = 4,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote14",
	filename = "__EmotesAndSpeechBubbles__/graphics/angry.png",
	size = 64,
	frame_count = 24,
	line_length = 4,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote15",
	filename = "__EmotesAndSpeechBubbles__/graphics/VibeCat.png",
	size = 64,
	frame_count = 159,
	line_length = 3,
	animation_speed = 0.4
},
{
	type = "animation",
	name = "emote16",
	filename = "__EmotesAndSpeechBubbles__/graphics/partyblob.png",
	size = 64,
	frame_count = 10,
	line_length = 5,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote17",
	filename = "__EmotesAndSpeechBubbles__/graphics/partyparrot.png",
	size = {64,46},
	frame_count = 10,
	line_length = 5,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote18",
	filename = "__EmotesAndSpeechBubbles__/graphics/hug.png",
	size = 64,
	frame_count = 76,
	line_length = 4,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote19",
	filename = "__EmotesAndSpeechBubbles__/graphics/bigbrain.png",
	size = 58,
	frame_count = 19,
	line_length = 1,
	scale = 1.25,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote20",
	filename = "__EmotesAndSpeechBubbles__/graphics/F.png",
	size = 112,
	frame_count = 8,
	line_length = 4,
	scale = 0.5,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote21",
	filename = "__EmotesAndSpeechBubbles__/graphics/smokeweedeveryday/rekt/rekt.png",
	size = 64,
	frame_count = 50,
	line_length = 5,
	animation_speed = 0.5,
	scale = 1.5
},
{
	type = "animation",
	name = "emote22",
	filename = "__EmotesAndSpeechBubbles__/graphics/coggers.png",
	size = 64,
	frame_count = 60,
	line_length = 5,
	animation_speed = 1
},
{
	type = "animation",
	name = "emote23",
	filename = "__EmotesAndSpeechBubbles__/graphics/disco.png",
	size = {228,128},
	tint = {1, 1, 1, 0.8},
	frame_count = 32,
	line_length = 1,
	animation_speed = 0.75,
	shift = {0, 1.5}
},
{
	type = "animation",
	name = "emote24",
	filename = "__EmotesAndSpeechBubbles__/graphics/stonks.png",
	size = {85,64},
	frame_count = 15,
	line_length = 5,
	animation_speed = 0.5
},
{
	type = "animation",
	name = "emote25",
	filename = "__EmotesAndSpeechBubbles__/graphics/catbeg.png",
	size = 64,
	frame_count = 45,
	line_length = 5,
	animation_speed = 0.25
},
{
	type = "animation",
	name = "emote26",
	filename = "__EmotesAndSpeechBubbles__/graphics/rockstare.png",
	size = 64,
	frame_count = 24,
	line_length = 6,
	animation_speed = 0.25
},
{
	type = "animation",
	name = "emote27",
	filename = "__EmotesAndSpeechBubbles__/graphics/catglare.png",
	size = {64,71},
	frame_count = 71,
	line_length = 1,
	animation_speed = 0.5
},
}

for each, emote in pairs(emotes) do
data:extend({
	emote,
	{
		type = "sprite",
		name = emote.name.."sprite",
		filename = emote.filename,
		priority = "extra-high",
		size = emote.size or math.min(emote.height, emote.width)
	},
	{ --------- The hatch recipie ----------
		type = "recipe",
		name = "ESB_EmoteCount",
		enabled = false,
		hidden = true,
		energy_required = #emotes+1,
		ingredients =
			{
			},
		result = "wooden-chest"
	},
})
end

data:extend({
  {
    type = "custom-input",
    name = "Emote",
    key_sequence = "mouse-button-3"
  },
  {
    type = "speech-bubble",
    name = "speech-bubble-no-fade",
    style = "compilatron_speech_bubble",
    wrapper_flow_style = "compilatron_speech_bubble_wrapper",
    fade_in_out_ticks = 5,
    flags = {"not-on-map", "placeable-off-grid"}
  }
})


control.lua
-- Setup tables and stuff for new/existing saves ----
script.on_init(
function(event)

if (global.AllPlayers == nil) then
	global.AllPlayers = {}
end

for PlayerID, PlayerLuaData in pairs(game.players) do
	if (global.AllPlayers[PlayerID] == nil) then
		global.AllPlayers[PlayerID] = {}
		global.AllPlayers[PlayerID].ChosenEmotes = {"emote2sprite", "emote4sprite", "emote5sprite", "emote8sprite", "emote11sprite", "emote12sprite", "emote16sprite", "emote19sprite", "emote20sprite"}
		global.AllPlayers[PlayerID].WheelSize = 45
		global.AllPlayers[PlayerID].EmoteDuration = 180
		global.AllPlayers[PlayerID].TextColor = {color="Player Color", index=1}
	end
end
end)

script.on_configuration_changed(
function(event)
if (global.AllPlayers == nil) then
	global.AllPlayers = {}
end

for PlayerID, PlayerLuaData in pairs(game.players) do
	if (global.AllPlayers[PlayerID] == nil) then
		global.AllPlayers[PlayerID] = {}
		global.AllPlayers[PlayerID].ChosenEmotes = {"emote2sprite", "emote4sprite", "emote5sprite", "emote8sprite", "emote11sprite", "emote12sprite", "emote16sprite", "emote19sprite", "emote20sprite"}
		global.AllPlayers[PlayerID].WheelSize = 45
		global.AllPlayers[PlayerID].EmoteDuration = 180
		global.AllPlayers[PlayerID].TextColor = {color="Player Color", index=1}
	end
end
end)

script.on_event(defines.events.on_player_created,
function(event)
if (global.AllPlayers[event.player_index] == nil) then
	global.AllPlayers[event.player_index] = {}
	global.AllPlayers[event.player_index].ChosenEmotes = {"emote2sprite", "emote4sprite", "emote5sprite", "emote8sprite", "emote11sprite", "emote12sprite", "emote16sprite", "emote19sprite", "emote20sprite"}
	global.AllPlayers[event.player_index].WheelSize = 45
	global.AllPlayers[event.player_index].EmoteDuration = 180
	global.AllPlayers[event.player_index].TextColor = {color="Player Color", index=1}
end
end)


script.on_event("Emote",
function(event) -- has .player_index, .input_name, and .cursor_position
local player = game.players[event.player_index]

if (player.opened_gui_type == defines.gui_type.none and not (player.gui.center["layout"] or player.gui.center["ChooseYourCharacter"])) then
player.gui.center.add{type="table", name="layout", column_count=3}
	player.gui.center.layout.add{type="table", name="words", column_count=1}
		player.gui.center.layout.words.add{type="label", name="LMB", caption = {"EmoteControls.LeftClick"}}
		player.gui.center.layout.words.add{type="label", name="RMB", caption = {"EmoteControls.RightClick"}}
		player.gui.center.layout.words.add{type="label", caption = {"EmoteControls.Size"}}
		player.gui.center.layout.words.add{type="button", name="SizeUp", caption="Larger"}
		player.gui.center.layout.words.add{type="button", name="SizeDown", caption="Smaller"}
		player.gui.center.layout.words.add{type="label", caption = {"EmoteControls.Duration"}, tooltip="99999 max"}
		player.gui.center.layout.words.add{type="textfield", name="EmoteDuration", numeric=true, allow_decimal=true, clear_and_focus_on_right_click=true, lose_focus_on_confirm=true, text=global.AllPlayers[event.player_index].EmoteDuration/60}
	player.gui.center.layout.add{type="table", name="buttons", column_count=3}
		for index, emote in pairs(global.AllPlayers[event.player_index].ChosenEmotes) do
		player.gui.center.layout.buttons.add{type="sprite-button", sprite=emote, name="b"..index, tags={spot = index}, tooltip = {"EmoteNames."..emote}}
		player.gui.center.layout.buttons["b"..index].style.size = global.AllPlayers[event.player_index].WheelSize
		end
	player.gui.center.layout.add{type="table", name="SBSettings", column_count=1}
		player.gui.center.layout.SBSettings.add{type="label", name="RMB", caption = "Speech Bubble Text Color"}
		player.gui.center.layout.SBSettings.add{type="drop-down", name="ColorOptions", items={"Player Color", "Default", "Red", "Orange", "Yellow", "Green", "Blue", "Purple", "Pink", "White", "Gray", "Black", "Brown", "Cyan", "Acid"}, selected_index=global.AllPlayers[event.player_index].TextColor.index}
elseif (player.gui.center["layout"]) then
	player.gui.center.layout.destroy()
end

end)

script.on_event(defines.events.on_gui_click,
--[[
element :: LuaGuiElement: The clicked element.
player_index :: uint: The player who did the clicking.
button :: defines.mouse_button_type: The mouse button used if any.
alt :: boolean: If alt was pressed.
control :: boolean: If control was pressed.
shift :: boolean: If shift was pressed.
]]
function(event)
local player = game.players[event.player_index]

if (player.character and event.element.parent.name == "buttons" and event.element.parent.parent.name == "layout") then
	if (event.button == defines.mouse_button_type.left) then
		if (global.AllPlayers[event.player_index].emote and rendering.is_valid(global.AllPlayers[event.player_index].emote)) then
			rendering.destroy(global.AllPlayers[event.player_index].emote)
		end
		global.AllPlayers[event.player_index].emote = rendering.draw_animation{
			animation = string.gsub(event.element.sprite, "sprite", ""),
			target = player.character,
			target_offset = {0, -3},
			surface = player.surface,
			animation_offset = 1,
			time_to_live = global.AllPlayers[event.player_index].EmoteDuration}
		global.AllPlayers[event.player_index].EmoteStartTick = game.tick
		player.gui.center.layout.destroy()
	elseif (event.button == defines.mouse_button_type.right) then
		global.AllPlayers[event.player_index].changing = event.element.tags.spot
		player.gui.center.layout.destroy()
		player.gui.center.add{type="table", name="ChooseYourCharacter", column_count=5}
			for i = 2, game.recipe_prototypes["ESB_EmoteCount"].energy do
			player.gui.center.ChooseYourCharacter.add{type="sprite-button", name="c"..i, sprite="emote"..i.."sprite", tooltip = {"EmoteNames.".."emote"..i.."sprite"}}
			player.gui.center.ChooseYourCharacter["c"..i].style.size = global.AllPlayers[event.player_index].WheelSize
			end
	end

elseif (player.character and event.element.parent.name == "ChooseYourCharacter" and event.button == defines.mouse_button_type.left) then
	-- for index, emote in pairs(global.AllPlayers[event.player_index].ChosenEmotes) do
		-- if (emote == global.AllPlayers[event.player_index].changing) then
			-- global.AllPlayers[event.player_index].ChosenEmotes[index] = event.element.sprite
		-- end
	-- end
	global.AllPlayers[event.player_index].ChosenEmotes[global.AllPlayers[event.player_index].changing] = event.element.sprite
	player.gui.center.ChooseYourCharacter.destroy()
	player.gui.center.add{type="table", name="layout", column_count=3}
	player.gui.center.layout.add{type="table", name="words", column_count=1}
		player.gui.center.layout.words.add{type="label", name="LMB", caption = {"EmoteControls.LeftClick"}}
		player.gui.center.layout.words.add{type="label", name="RMB", caption = {"EmoteControls.RightClick"}}
		player.gui.center.layout.words.add{type="label", caption = {"EmoteControls.Size"}}
		player.gui.center.layout.words.add{type="button", name="SizeUp", caption="Larger"}
		player.gui.center.layout.words.add{type="button", name="SizeDown", caption="Smaller"}
		player.gui.center.layout.words.add{type="label", caption = {"EmoteControls.Duration"}, tooltip="99999 max"}
		player.gui.center.layout.words.add{type="textfield", name="EmoteDuration", numeric=true, allow_decimal=true, clear_and_focus_on_right_click=true, lose_focus_on_confirm=true, text=global.AllPlayers[event.player_index].EmoteDuration/60}
	player.gui.center.layout.add{type="table", name="buttons", column_count=3}
		for index, emote in pairs(global.AllPlayers[event.player_index].ChosenEmotes) do
		player.gui.center.layout.buttons.add{type="sprite-button", sprite=emote, name="b"..index, tags={spot = index}, tooltip = {"EmoteNames."..emote}}
		player.gui.center.layout.buttons["b"..index].style.size = global.AllPlayers[event.player_index].WheelSize
		end
	player.gui.center.layout.add{type="table", name="SBSettings", column_count=1}
		player.gui.center.layout.SBSettings.add{type="label", name="RMB", caption = "Speech Bubble Text Color"}
		player.gui.center.layout.SBSettings.add{type="drop-down", name="ColorOptions", items={"Player Color", "Default", "Red", "Orange", "Yellow", "Green", "Blue", "Purple", "Pink", "White", "Gray", "Black", "Brown", "Cyan", "Acid"}, selected_index=global.AllPlayers[event.player_index].TextColor.index}
elseif (event.element.name == "SizeUp" and event.element.parent.name == "words") then
	global.AllPlayers[event.player_index].WheelSize = global.AllPlayers[event.player_index].WheelSize+10
	for index, emote in pairs(global.AllPlayers[event.player_index].ChosenEmotes) do
		player.gui.center.layout.buttons["b"..index].style.size = global.AllPlayers[event.player_index].WheelSize
	end

elseif (event.element.name == "SizeDown" and event.element.parent.name == "words" and global.AllPlayers[event.player_index].WheelSize > 10) then
	global.AllPlayers[event.player_index].WheelSize = global.AllPlayers[event.player_index].WheelSize-10
	for index, emote in pairs(global.AllPlayers[event.player_index].ChosenEmotes) do
		player.gui.center.layout.buttons["b"..index].style.size = global.AllPlayers[event.player_index].WheelSize
	end

end

end)


script.on_event(defines.events.on_gui_selection_state_changed,
-- element :: LuaGuiElement: The element whose value changed.
-- player_index :: uint: The player who did the change.
function(event)
local player = game.players[event.player_index]
if (event.element.name == "ColorOptions" and event.element.parent.name == "SBSettings") then
	global.AllPlayers[event.player_index].TextColor = {color=event.element.get_item(event.element.selected_index), index=event.element.selected_index}
end
end)

-- script.on_event(defines.events.on_gui_text_changed,
-- -- element :: LuaGuiElement: The edited element.
-- -- player_index :: uint: The player who did the edit.
-- -- text :: string: The new text in the element.
-- -- name :: defines.events: Identifier of the event
-- -- tick :: uint: Tick the event was generated.
-- function(event)
-- if (event.element.name == "EmoteDuration" and event.element.parent.name == "words") then
-- 	global.AllPlayers[event.player_index].EmoteDuration = event.text*60
-- end
-- end)

script.on_event(defines.events.on_gui_confirmed,
function(event)
if (event.element.name == "EmoteDuration" and event.element.parent.name == "words" and event.element.text ~= "" and event.element.text ~= "0" and tonumber(event.element.text) <= 99999) then
	global.AllPlayers[event.player_index].EmoteDuration = event.element.text*60
	if (global.AllPlayers[event.player_index].emote and rendering.is_valid(global.AllPlayers[event.player_index].emote)) then
		rendering.destroy(global.AllPlayers[event.player_index].emote)
	end
	global.AllPlayers[event.player_index].emote = nil
end
end)

script.on_event(defines.events.on_tick,
function(event)
for the, properties in pairs(global.AllPlayers) do
	if (properties.emote and rendering.is_valid(properties.emote) and game.tick < properties.EmoteStartTick+10) then
		rendering.set_target(properties.emote, game.players[the].character, {0, -0.3-(game.tick-properties.EmoteStartTick)/3})
		rendering.set_x_scale(properties.emote, (game.tick+1-properties.EmoteStartTick)/10)
		rendering.set_y_scale(properties.emote, (game.tick+1-properties.EmoteStartTick)/10)
	elseif (properties.emote and rendering.is_valid(properties.emote) and game.tick > properties.EmoteStartTick+(properties.EmoteDuration-10)) then
		rendering.set_target(properties.emote, game.players[the].character, {0, -3.6333+((game.tick-(properties.EmoteStartTick+(properties.EmoteDuration-10)))/3)})
		rendering.set_x_scale(properties.emote, 1-(game.tick-(properties.EmoteStartTick+(properties.EmoteDuration-10)))/10)
		rendering.set_y_scale(properties.emote, 1-(game.tick-(properties.EmoteStartTick+(properties.EmoteDuration-10)))/10)
	end
end
end)

script.on_event(defines.events.on_console_chat,
-- player_index :: uint (optional): The player if any.
-- message :: string: The chat message.
function(event)
if (event.player_index) then
	local player = game.players[event.player_index]
	if (player.character) then
		if (global.AllPlayers[event.player_index].bubble and global.AllPlayers[event.player_index].bubble.valid) then
			global.AllPlayers[event.player_index].bubble.destroy()
		end
		local ColorBracket = ""
		if (global.AllPlayers[event.player_index].TextColor.color == "Player Color") then
			ColorBracket = "[color="..player.color.r..","..player.color.g..","..player.color.b.."]"
		else
			ColorBracket = "[color="..string.lower(global.AllPlayers[event.player_index].TextColor.color).."]"
		end
		global.AllPlayers[event.player_index].bubble = player.surface.create_entity
			{
				name = "speech-bubble-no-fade",
				position = player.character.position,
				source = player.character,
				text = ColorBracket..event.message.."[/color]",
				lifetime = 240
			}
	end
end
end)
